src/kbmod/work_unit.py

```diff
@@ -541,6 +548,237 @@ def to_fits(self, filename, overwrite=False):
+
+        # Set up the initial HDU list, including the primary header
+        # the metadata (empty), and the configuration.
+        hdul = fits.HDUList()
+        pri = fits.PrimaryHDU()
+        pri.header["NUMIMG"] = self.im_stack.img_count()
+        pri.header["REPRJCTD"] = self.reprojected
+        pri.header["HELIO"] = self.heliocentric_distance
+        for i in range(len(self.constituent_images)):
+            pri.header[f"GEO_{i}"] = self.geocentric_distances[i]
+
+        # If the global WCS exists, append the corresponding keys.
+        if self.wcs is not None:
+            append_wcs_to_hdu_header(self.wcs, pri.header)
+
+        n_constituents = len(self.constituent_images)
+        pri.header["NCON"] = n_constituents
+
+        hdul.append(pri)
+
+        meta_hdu = fits.BinTableHDU()
+        meta_hdu.name = "metadata"
+        hdul.append(meta_hdu)
+
+        config_hdu = self.config.to_hdu()
+        config_hdu.name = "kbmod_config"
+        hdul.append(config_hdu)
+
+        for i in range(self.im_stack.img_count()):
+            layered = self.im_stack.get_single_image(i)
+            c_indices = self._per_image_indices[i]
+            n_indices = len(c_indices)
+            sub_hdul = fits.HDUList()
+
+            img_wcs = self.get_wcs(i)
+            sci_hdu = raw_image_to_hdu(layered.get_science(), img_wcs)
+            sci_hdu.name = f"SCI_{i}"
+            sci_hdu.header["NIND"] = n_indices
+            for j in range(n_indices):
+                sci_hdu.header[f"IND_{j}"] = c_indices[j]
+            sub_hdul.append(sci_hdu)
+
+            var_hdu = raw_image_to_hdu(layered.get_variance())
+            var_hdu.name = f"VAR_{i}"
+            sub_hdul.append(var_hdu)
+
+            msk_hdu = raw_image_to_hdu(layered.get_mask())
+            msk_hdu.name = f"MSK_{i}"
+            sub_hdul.append(msk_hdu)
+
+            p = layered.get_psf()
+            psf_array = np.array(p.get_kernel()).reshape((p.get_dim(), p.get_dim()))
+            psf_hdu = fits.hdu.image.ImageHDU(psf_array)
+            psf_hdu.name = f"PSF_{i}"
+            sub_hdul.append(psf_hdu)
+            sub_hdul.writeto(os.path.join(directory, f"{i}_{filename}"))
+
+        for i in range(n_constituents):
+            img_location = self.constituent_images[i]
+
+            orig_wcs = self._per_image_wcs[i]
+            wcs_hdu = fits.TableHDU()
+            append_wcs_to_hdu_header(orig_wcs, wcs_hdu.header)
+            wcs_hdu.name = f"WCS_{i}"
+            wcs_hdu.header["ILOC"] = img_location
+            hdul.append(wcs_hdu)
+
+            im_ebd_wcs = self._per_image_ebd_wcs[i]
+            ebd_hdu = fits.TableHDU()
+            append_wcs_to_hdu_header(im_ebd_wcs, ebd_hdu.header)
+            ebd_hdu.name = f"EBD_{i}"
+            hdul.append(ebd_hdu)
+
+        hdul.writeto(os.path.join(directory, filename), overwrite=overwrite)
+
+    @classmethod
+    def from_fits_shard(cls, filename, directory, lazy=False):
```

Comment:
As a minor naming note, I would suggest `from_sharded_fits`. But I'm fine if you want to keep it as-is.